### PR TITLE
Extend `#[must_use]` to nested structures

### LINKED ISSUE #39524: #[must_use] does not work for nested structures
When `#[must_use]` is applied to a structure that is nested, the linter no longer picks up when that value is unused.

This is related to: #26291 and #26281

Looking at those issues, there seem to be a lot of cases where `#[must_use]` does not work. This is unfortunate because it is actually an incredibly useful annotation that definitely has applications past just the Result enum. 

Accidentally relying on it too much can have you make mistakes you otherwise would have corrected had the warning been a bit smarter. I consider this a bug because `#[must_use]` implies that you must use that value no matter what; regardless of what form it is returned in.

I tried this code: (Rust Playground: https://is.gd/vCLZKq)

```rust
#[must_use]
#[derive(Debug)]
struct MustUse;

struct Foo {
    must: MustUse,
}

impl Foo {
    fn bar(&self) {
        println!("{:?}", self.must);
    }
}

fn must() -> Foo {
    Foo {must: MustUse}
}

fn main() {
    must().bar();
    must();
}
```

I expected to see this happen: A warning about the MustUse structure not being used

Instead, this happened: The code compiled and ran with no errors

## Meta

`rustc --version --verbose`: stable, nightly, etc.

Backtrace: not applicable

### PR DESCRIPTION
**Pull request by varkor**
Extends the `#[must_use]` lint to apply when `#[must_use]` types are nested within `struct`s (or one-variant `enum`s), making the lint much more generally useful. This is in line with https://github.com/rust-lang/rust/pull/61100 extending the lint to tuples.

Fixes https://github.com/rust-lang/rust/issues/39524.

cc @rust-lang/lang and @rust-lang/compiler for discussion in case this is a controversial change. In particular, we might want to consider allowing annotations on fields containing `#[must_use]` types in user-defined types (e.g. `#[allow(unused_must_use)]`) to opt out of this behaviour, if there are cases where we this this is likely to have frequent false positives.

(This is based on top of https://github.com/rust-lang/rust/pull/62235.)

### CODE REVIEW

File: src/librustc_lint/unused.rs [hunk 1]
```diff
@@ -151,8 +154,40 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedResults {
...
-                    check_must_use_def(cx, def.did, span, descr_pre, descr_post)
+                ty::Adt(def, subst) => {
+                    // Check the type itself for `#[must_use]` annotations.
+                    let mut has_emitted = check_must_use_def(
+                        cx, def.did, span, descr_pre, descr_post);
+                    // Check any fields of the type for `#[must_use]` annotations.
+                    // We ignore ADTs with more than one variant for simplicity and to avoid
+                    // false positives.
+                    // Unions are also ignored (though in theory, we could lint if every field of
+                    // a union was `#[must_use]`).
```
Comment (Centril) [inline src/librustc_lint/unused.rs]: This should probably go into a separate method and the comments above can become doc comments...

File: src/librustc_lint/unused.rs [hunk 2]
```diff
@@ -202,24 +237,42 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedResults {
...
-                    // If the array is definitely non-empty, we can do `#[must_use]` checking.
-                    Some(n) if n != 0 => {
-                        let descr_pre = &format!(
-                            "{}array{} of ",
-                            descr_pre,
-                            plural_suffix,
-                        );
-                        check_must_use_ty(cx, ty, expr, span, descr_pre, descr_post, true)
+                ty::Array(ty, mut len) => {
+                    // Try to evaluate the length if it's unevaluated.
```
Comment (Centril) [inline src/librustc_lint/unused.rs]: It feels like there should be some function that already does this that you can reuse?

**Comment by estebank** [comment]
I'm not available for the next couple of days due to personal reasons. I would appreciate it if someone else could take over review.

**Comment by Mark-Simulacrum** [comment]
I don't think the review is in any rush - this definitely need lang-team sign off (and maybe libs?); I feel like we've discussed this previously and concluded that we shouldn't do it, IIRC because it expands it far enough that it feels like it "always" works which isn't actually the case. I could be projecting my own thoughts on the matter though.

**Comment by Centril** [comment]
> I don't think the review is in any rush - this definitely need lang-team sign off (and maybe libs?);

I nominated this for the lang team. The libs team doesn't deal with compiler lints.

----------

I personally feel that the false-positive cases are edge cases that are rare enough that they don't outweigh the big benefit this PR brings.

We could however tweak this a bit to look for a `Drop` implementations for the ADT and then bail on the structural logic tho that might be too subtle. An opt-out would also be decent.

**Comment by Mark-Simulacrum** [comment]
I disagree a little bit about the libs team bit here -- it seems like this lint is predominantly used for libraries, not language features/things, and so their opinion matters more. At the very least, they should be notified (cc @rust-lang/libs).

File: src/libstd/panicking.rs [hunk 3]
```diff
@@ -103,7 +103,9 @@ pub fn set_hook(hook: Box<dyn Fn(&PanicInfo<'_>) + 'static + Sync + Send>) {
         HOOK_LOCK.write_unlock();
 
         if let Hook::Custom(ptr) = old_hook {
-            Box::from_raw(ptr);
+            #[allow(unused_must_use)] {
+                Box::from_raw(ptr);
```
Comment (eddyb) [inline src/libstd/panicking.rs]: This could be `mem::drop(Box::from_raw(ptr));` for clarity.

Comment (eddyb) [reply src/libstd/panicking.rs]: `let _ = ...;` would also work though, surprised you didn't go with that?

File: src/librustc_lint/unused.rs [hunk 4]
```diff
@@ -179,14 +235,45 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedResults {
...
+                        // If the array is definitely non-empty, we can do `#[must_use]` checking.
+                        Some(n) if n != 0 => {
+                            let descr_pre = &format!(
+                                "{}array{} of ",
+                                descr_pre,
+                                plural_suffix,
+                            );
+                            check_must_use_ty(cx, ty, expr, span, descr_pre, descr_post, true)
+                        }
+                        // Otherwise, we don't lint, to avoid false positives.
```
Comment (eddyb) [inline src/librustc_lint/unused.rs]: I would do this the other way around: always lint unless *definitely* `0`.
Especially if it's generic, that would mean it's *sometimes* `0` (or never, 0-length arrays are rare IMO).

File: src/librustc_lint/unused.rs [hunk 5]
```diff
@@ -133,23 +136,71 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedResults {
...
+                ty::Adt(def, subst) => {
+                    // Check the type itself for `#[must_use]` annotations.
+                    let mut has_emitted = check_must_use_def(
+                        cx, def.did, span, descr_pre, descr_post);
+                    // Check any fields of the type for `#[must_use]` annotations.
+                    // We ignore ADTs with more than one variant for simplicity and to avoid
+                    // false positives.
+                    // Unions are also ignored (though in theory, we could lint if every field of
+                    // a union was `#[must_use]`).
+                    if def.variants.len() == 1 && !def.is_union() {
```
Comment (eddyb) [inline src/librustc_lint/unused.rs]: If there are *any* variants that are `#[must_use]`, the whole `enum` probably should be too (but you'd need to prevent infinite recursion somehow).

Comment (eddyb) [reply src/librustc_lint/unused.rs]: Maybe we should switch to a query, like the `sized_requirements_for_adt` or whatever that's called?

**Comment by Nemo157** [comment]
Would it make sense to do a crater run of this to see how useful the warnings it produces in real code are?

**Comment by nikomatsakis** [comment]
Discussed briefly in @rust-lang/lang -- we agree a crater run would be the logical next step before making any further decisions. Can we set this to deny by default and do a run?

File: src/librustc_lint/unused.rs (hunk 2 above)
Comment (varkor) [reply src/librustc_lint/unused.rs]: Once https://github.com/rust-lang/rust/pull/59369 is merged, this should become unnecessary. I'll add a `FIXME` regarding it.

File: src/librustc_lint/unused.rs (hunk 5 above)
Comment (varkor) [reply src/librustc_lint/unused.rs]: I was thinking of warning if there are any `#[must_use]` variants, but I thought it could get quite noisy with false positives. If https://github.com/rust-lang/rfcs/pull/2593 or similar was merged, then we could do this with fewer false positives.

**Comment by estebank** [comment]
Minor test fixing needed:

```
---- [run-pass] run-pass/cross-crate/cci_capture_clause.rs stdout ----

error: auxiliary build of "/checkout/src/test/run-pass/cross-crate/auxiliary/cci_capture_clause.rs" failed to compile: 
status: exit code: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/run-pass/cross-crate/auxiliary/cci_capture_clause.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/cross-crate/cci_capture_clause/auxiliary" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--crate-type" "dylib" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/cross-crate/cci_capture_clause/auxiliary"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
error: unused `std::result::Result` that must be used
  --> /checkout/src/test/run-pass/cross-crate/auxiliary/cci_capture_clause.rs:7:9
   |
LL |         tx.send(x.clone());
   |         ^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[deny(unused_must_use)]` on by default
   = note: this `Result` may be an `Err` variant, which should be handled

error: aborting due to previous error


------------------------------------------


---- [run-pass] run-pass/issues/issue-2723-b.rs stdout ----

error: auxiliary build of "/checkout/src/test/run-pass/issues/auxiliary/issue-2723-a.rs" failed to compile: 
status: exit code: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/run-pass/issues/auxiliary/issue-2723-a.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/issues/issue-2723-b/auxiliary" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--crate-type" "dylib" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/issues/issue-2723-b/auxiliary"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
warning: function is never used: `q`
  --> /checkout/src/test/run-pass/issues/auxiliary/issue-2723-a.rs:2:26
   |
LL |     xs.iter().map(|_x| { unsafe fn q() { panic!(); } }).collect::<Vec<()>>();
   |                          ^^^^^^^^^^^^^
   |
   = note: `#[warn(dead_code)]` on by default

error: unused return value of `std::iter::Iterator::collect` that must be used
  --> /checkout/src/test/run-pass/issues/auxiliary/issue-2723-a.rs:2:5
   |
LL |     xs.iter().map(|_x| { unsafe fn q() { panic!(); } }).collect::<Vec<()>>();
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[deny(unused_must_use)]` on by default
   = note: if you really need to exhaust the iterator, consider `.for_each(drop)` instead

error: aborting due to previous error


------------------------------------------


---- [run-pass] run-pass/issues/issue-9906.rs stdout ----

error: auxiliary build of "/checkout/src/test/run-pass/issues/auxiliary/issue-9906.rs" failed to compile: 
status: exit code: 1
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/run-pass/issues/auxiliary/issue-9906.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/issues/issue-9906/auxiliary" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--crate-type" "dylib" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-pass/issues/issue-9906/auxiliary"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
warning: field is never used: `value`
  --> /checkout/src/test/run-pass/issues/auxiliary/issue-9906.rs:5:23
   |
LL |     pub struct FooBar{value: isize}
   |                       ^^^^^^^^^^^^
   |
   = note: `#[warn(dead_code)]` on by default

error: unused arithmetic operation that must be used
  --> /checkout/src/test/run-pass/issues/auxiliary/issue-9906.rs:13:9
   |
LL |         1+1;
   |         ^^^
   |
   = note: `#[deny(unused_must_use)]` on by default

error: aborting due to previous error


------------------------------------------



failures:
    [run-pass] run-pass/cross-crate/cci_capture_clause.rs
    [run-pass] run-pass/issues/issue-2723-b.rs
    [run-pass] run-pass/issues/issue-9906.rs
```

File: src/librustc_lint/unused.rs [hunk 6]
```diff
@@ -23,7 +26,7 @@ use log::debug;
 
 declare_lint! {
     pub UNUSED_MUST_USE,
-    Warn,
+    Deny,
```
Comment (estebank) [inline src/librustc_lint/unused.rs]: Should we make this deny by default in a separate PR? I'm a bit weary of simultaneously making this broader and more constraining (unless this is just for crater's sake, in that case carry right along).

**Review by estebank — APPROVED** [review]

File: src/librustc_lint/unused.rs (hunk 6 above)
Comment (Centril) [reply src/librustc_lint/unused.rs]: That's just for crater purposes.

**Review by estebank — APPROVED** [review]

**Comment by gagan0723** [comment]
Ping from triage, @varkor any update on this?

Some tests are failing and this PR has merge conflicts that needs to be resolved.

**Comment by varkor** [comment]
I'm meaning to get back to this, but it may take me a couple of weeks.

File: src/test/pretty/unary-op-disambig.rs [hunk 7]
```diff
@@ -16,4 +16,4 @@ fn alt_semi() -> isize { match true { true => { f() } _ => { } }; -1 }
 
 fn alt_no_semi() -> isize { (match true { true => { 0 } _ => { 1 } }) - 1 }
 
-fn stmt() { { f() }; -1; }
+fn stmt() { { f() }; let _ = -1; }
```
Comment (estebank) [inline src/test/pretty/unary-op-disambig.rs]: Isn't this _not_ testing what the test is trying to test? I don't see this ever regressing, but it is something to keep in mind.

File: src/test/pretty/block-disambig.rs [hunk 8]
```diff
@@ -7,7 +7,7 @@
 
 use std::cell::Cell;
 
-fn test1() { let val = &0; { } *val; }
+fn test1() { let val = &0; { } let _ = *val; }
```
Comment (estebank) [inline src/test/pretty/block-disambig.rs]: This is an explicit test for that block not to be identified as multiplied by `val`. Disable the lint here.

**Comment by JohnCSimon** [comment]
Ping from triage 
@varkor Can you please resolve the merge conflicts? Thanks

**Comment by varkor** [comment]
I've opened https://github.com/rust-lang/cargo/pull/7340 so that cargo will compile with these changes and we can hopefully carry out a crater run.

**Comment by alexcrichton** [comment]
I wanted to comment here a bit before https://github.com/rust-lang/cargo/pull/7340 is merged because this looks sort of odd to me, it looks like the error on CI is:

```
error: unused boxed boxed `std::ops::FnMut` trait object in field `socket` that must be used
  --> src/tools/cargo/src/cargo/ops/cargo_package.rs:47:9
   |
47 |         ops::resolve_ws(ws)?;
   |         ^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[deny(unused_must_use)]` on by default
   = note: closures are lazy and do nothing unless called

error: unused boxed boxed `std::ops::FnMut` trait object in field `timer` that must be used
  --> src/tools/cargo/src/cargo/ops/cargo_package.rs:47:9
   |
47 |         ops::resolve_ws(ws)?;
   |         ^^^^^^^^^^^^^^^^^^^^^
   |
   = note: closures are lazy and do nothing unless called
```

That looks pretty weird because there's no field `socket` nor `timer` anywhere in the Cargo codebase. This to me seems like pretty bad UX and I wouldn't really want to have to deal with warnings like this in Cargo. 

It looks like this is still in the evaluation phase? Does this just want to run through crater to see what the impact is? I would prefer that this not land precisely in this form, even if crater is clean-ish because this is such an inscrutable warning.

**Comment by varkor** [comment]
Yes, I was intending to clean up the diagnostics after the crater run (I didn't want to spend too much time on them now in case the changes were simply deemed infeasible) — I agree that those errors aren't helpful from a UX perspective.

**Comment by varkor** [comment]
@craterbot run mode=check-only

**Comment by pietroalbini** [comment]
@craterbot requirement=linux

**Comment by Centril** [comment]
A quick skimming suggests that this is catching some bugs.

@varkor can you distill this into a report and then I-nominate for the lang team once ready?

**Comment by varkor** [comment]
Quick report:

Most of the "regressions" are actually just arising from making `#[must_use]` a deny-by-default lint, where people were already ignoring the warning, and result from ignored `Result` types. Maybe 1 in 20 actually has new behaviour after this PR. Examples: [1](https://crater-reports.s3.amazonaws.com/pr-62262/try%235ac1de79c89dcd229444180c480a54cc60aa28ae/gh/Co0sh.hexagon-engine/log.txt), [2](https://crater-reports.s3.amazonaws.com/pr-62262/try%235ac1de79c89dcd229444180c480a54cc60aa28ae/gh/Byron.git-reconstruct/log.txt), [3](https://crater-reports.s3.amazonaws.com/pr-62262/try%235ac1de79c89dcd229444180c480a54cc60aa28ae/gh/CMatri.tethermanager/log.txt), [4](https://crater-reports.s3.amazonaws.com/pr-62262/try%235ac1de79c89dcd229444180c480a54cc60aa28ae/gh/CallumJHays.wasm-bindgen-futures-test/log.txt), [5](https://crater-reports.s3.amazonaws.com/pr-62262/try%235ac1de79c89dcd229444180c480a54cc60aa28ae/gh/Baoyx007.pna/log.txt).
As far as I can tell, these are bugs, and come from overlooking nested `#[must_use]` types (often `Result`s), and so extending this lint would be worthwhile for those users.

I personally think we should extend the lint to cover these nested uses (though we should either wait until the diagnostics have been improved following [@alexcrichton's comment](https://github.com/rust-lang/rust/pull/62262#issuecomment-529514063) before r+ing, or file a follow up PR to this one addressing the concern).

**Comment by Nemo157** [comment]
That [last example](https://crater-reports.s3.amazonaws.com/pr-62262/try%235ac1de79c89dcd229444180c480a54cc60aa28ae/gh/Baoyx007.pna/log.txt) is interesting as it's triggering the warning 3 times for the same value. It seems like fields of `#[must_use]` types should be ignored and only the warning from the outer-most `must_use` should trigger.

**Comment by withoutboats** [comment]
> https://crater-reports.s3.amazonaws.com/pr-62262/try%235ac1de79c89dcd229444180c480a54cc60aa28ae/gh/Co0sh.hexagon-engine/log.txt

This example is pretty instructive: currently the Result is used in all of these futures in an internal way, and these futures should not be treated as `must_use` - dropping futures to cancel them is an integral part of how futures work. Making every cancellation of a future into a bug would be a very disruptive change.

Discussing this in the lang team meeting, we reached the conclusion that this is basically a disruptive idiom shift along the same lines of many of the edition 2018 changes, like `dyn Trait`. Currently, the default assumption is that you can put results inside your types, and its *your* responsibility as the type's definer to mark it `must_use`. This shifts the burden in the opposite direction - if your type stores a must_use type but isn't must use, you are responsible for annotating it.

We also noted some variants which would probably be less disruptive and more narrowly targeted:

- When a field which is a type parameter is a must_use (e.g. `Option<Result<..>>` is must_use)
- When a field which is public is must_use

Basically, moving forward with this I think would require an RFC, justifying the change, documenting the change in usage patterns, idioms, etc. We need to involve the rest of the community in the discussion so that the change is more widely known and agreed upon before we move forward with changes falling into this sort of pattern.

**Comment by Nemo157** [comment]
> > https://crater-reports.s3.amazonaws.com/pr-62262/try%235ac1de79c89dcd229444180c480a54cc60aa28ae/gh/Co0sh.hexagon-engine/log.txt
> 
> This example is pretty instructive: currently the Result is used in all of these futures in an internal way, and these futures should not be treated as `must_use` - dropping futures to cancel them is an integral part of how futures work. Making every cancellation of a future into a bug would be a very disruptive change.

That doesn't appear to be what this log is showing. The error messages are coming from the line

```rust
runtime.block_on(runner)?;
```

because the result of this expression has type `(futures::stream::Select<_>, _)`. If `futures::stream::Forward` had happened to only return the stream and not a (stream, sink) tuple this would already have been warning about the dropped `#[must_use]` type. IMO this seems to be the most uncontroversial of the extensions, returning a `#[must_use]` value and another value in a tuple should suppress the `#[must_use]` lint.

In fact, compiling `hexagon-engine` with `rustc 1.38.0 (625451e37 2019-09-23)` I already get the base warning (and the other warnings emitted would be suppressed by my suggestion above to not re-warn about the same value):

```
warning: unused `futures::stream::Select` in tuple element 0 that must be used
  --> src/connect.rs:43:5
   |
43 |     runtime.block_on(runner)?;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_must_use)]` on by default
   = note: streams do nothing unless polled
```
(I don't deny that there might be some bad interactions with some form of select-for-cancellation, but because of futures 0.1's APIs I doubt those are going to be visible in existing crates, this seems like something that needs a specific example written and checked against this change to see what the interaction is.)

**Comment by withoutboats** [comment]
> because the result of this expression has type (futures::stream::Select<_>, _). If futures::stream::Forward had happened to only return the stream and not a (stream, sink) tuple this would already have been warning about the dropped #[must_use] type.

I see that these types *were* tagged must_use in futures 0.1. But I'm a bit confused by your comment as well, because in the lang team meeting @cramertj said that we have already landed the change to apply must_use to tuples.

**Comment by cramertj** [comment]
Yup, that was reported in https://github.com/rust-lang/rust/issues/61061 and fixed in https://github.com/rust-lang/rust/pull/61100.

**Comment by Nemo157** [comment]
Ah, yeah, I should have edited that bit of my comment, I thought that change was still nightly-only when I wrote that paragraph, I went and tested it on stable afterwards and forgot to change it.

So `hexagon-engine` isn't fundamentally affected by this change, it's already warning on stable for this value.

**Comment by varkor** [comment]
> When a field which is public is `must_use`

I think this was really an oversight on my part, and we should just be checking those fields that are public. Internal details of the types should be hidden. If I restricted this PR to just those `pub` fields, would that be sufficient to go ahead (perhaps subject to another crater run), or would an RFC still be necessary?

**Comment by Centril** [comment]
> (perhaps subject to another crater run)

We have two agents at the moment so at least we have the resources for this now.

Please make sure you don't inflate the count this time around tho =)

**Comment by varkor** [comment]
@craterbot check

**Comment by varkor** [comment]
23 crates produced the new lint (most of the regressions caught by crater were downstream). They all seem genuine to me, though in most cases they were deliberately (though not explicitly) ignored (in things like `drop` impls, etc.). In these cases, `let _ = ...;` should be appropriate.

I think this crater run demonstrates the change is not very invasive, but can catch genuine errors, and is worth merging.

- [arraydeque](https://crater-reports.s3.amazonaws.com/pr-62262-1/try%23ea663bba38739867a4b75ac820991b4f5d093c3b/gh/0xpr03.service-daemon/log.txt)
- [midir](https://crater-reports.s3.amazonaws.com/pr-62262-1/try%23ea663bba38739867a4b75ac820991b4f5d093c3b/gh/ColinKinloch.lpd8_config/log.txt)
- [web-view](https://crater-reports.s3.amazonaws.com/pr-62262-1/try%23ea663bba38739867a4b75ac820991b4f5d093c3b/gh/JackMordaunt.nativefier/log.txt)
- [safe-graph](https://crater-reports.s3.amazonaws.com/pr-62262-1/try%23ea663bba38739867a4b75ac820991b4f5d093c3b/reg/safe-graph-0.1.6/log.txt)
- [sodium](https://crater-reports.s3.amazonaws.com/pr-62262-1/try%23ea663bba38739867a4b75ac820991b4f5d093c3b/gh/clinuxrulz.sodium-rust-push-pull/log.txt)
- [renamefs](https://crater-reports.s3.amazonaws.com/pr-62262-1/try%23ea663bba38739867a4b75ac820991b4f5d093c3b/gh/comicfans.renamefs/log.txt)
- [geniza](https://crater-reports.s3.amazonaws.com/pr-62262-1/try%23ea663bba38739867a4b75ac820991b4f5d093c3b/gh/bnewbold.geniza/log.txt)
- [tetris_piston](https://crater-reports.s3.amazonaws.com/pr-62262-1/try%23ea663bba38739867a4b75ac820991b4f5d093c3b/gh/dunnker.tetris-piston/log.txt)
- [prometrics](https://crater-reports.s3.amazonaws.com/pr-62262-1/try%23ea663bba38739867a4b75ac820991b4f5d093c3b/gh/frugalos.watarikani/log.txt)
- [synchronize_images](https://crater-reports.s3.amazonaws.com/pr-62262-1/try%23ea663bba38739867a4b75ac820991b4f5d093c3b/gh/gadomski.synchronize-images/log.txt)
- [corona](https://crater-reports.s3.amazonaws.com/pr-62262-1/try%23ea663bba38739867a4b75ac820991b4f5d093c3b/gh/vorner.corona/log.txt)
- [lingo_proxy](https://crater-reports.s3.amazonaws.com/pr-62262-1/try%23ea663bba38739867a4b75ac820991b4f5d093c3b/gh/wayslog.lingo-proxy/log.txt)
- [tox](https://crater-reports.s3.amazonaws.com/pr-62262-1/try%23ea663bba38739867a4b75ac820991b4f5d093c3b/gh/zetok.tox/log.txt)
- [tox](https://crater-reports.s3.amazonaws.com/pr-62262-1/try%23ea663bba38739867a4b75ac820991b4f5d093c3b/reg/tox-0.0.10/log.txt)
- [connetr](https://crater-reports.s3.amazonaws.com/pr-62262-1/try%23ea663bba38739867a4b75ac820991b4f5d093c3b/reg/connectr-1.1.0/log.txt)
- [runtime](https://crater-reports.s3.amazonaws.com/pr-62262-1/try%23ea663bba38739867a4b75ac820991b4f5d093c3b/reg/runtime-0.3.0-alpha.7/log.txt)
- [slice-deque](https://crater-reports.s3.amazonaws.com/pr-62262-1/try%23ea663bba38739867a4b75ac820991b4f5d093c3b/reg/slice-deque-0.2.4/log.txt)
- [stdweb](https://crater-reports.s3.amazonaws.com/pr-62262-1/try%23ea663bba38739867a4b75ac820991b4f5d093c3b/reg/stdweb-0.4.18/log.txt)
- [frappe](https://crater-reports.s3.amazonaws.com/pr-62262-1/try%23ea663bba38739867a4b75ac820991b4f5d093c3b/reg/ted-tui-0.5.0/log.txt)
- [botana-api-rs]( https://crater-reports.s3.amazonaws.com/pr-62262-1/try%23ea663bba38739867a4b75ac820991b4f5d093c3b/gh/botnana.botnana-api-rs/log.txt)
- [lifegame](https://crater-reports.s3.amazonaws.com/pr-62262-1/try%23ea663bba38739867a4b75ac820991b4f5d093c3b/gh/taturou.lifegame-core/log.txt)
- [mdl](https://crater-reports.s3.amazonaws.com/pr-62262-1/try%23ea663bba38739867a4b75ac820991b4f5d093c3b/reg/mdl-1.0.4/log.txt)
- [sodium-rust](https://crater-reports.s3.amazonaws.com/pr-62262-1/try%23ea663bba38739867a4b75ac820991b4f5d093c3b/reg/sodium-rust-1.0.1/log.txt)

**Comment by estebank** [comment]
The one case I noticed that might be reasonable to allow is calling `drain`.

**Comment by varkor** [comment]
I don't think that pattern is common: the only reason that instance was linted is because they were implementing `clear` using `drain(..)`. It seems reasonable to want `let _ = self.drain(..);` in this case.

**Comment by varkor** [comment]
Nominating for lang team re-review.

**Comment by cramertj** [comment]
We discussed this in the lang meeting today and there was general agreement that the lint as-implemented in this PR isn't something we're interested in landing. This change would make `#[must_use]` operate similarly to an auto-trait in that `#[must_used]`-ness would be inherited from fields, but with no ability to turn off the `#[must_used]`-ness of a type which contains a `#[must_use]` type. A number of the examples in the crater run above (e.g. `Drain`) demonstrate that this is not what authors intend by `#[must_use]`, and in many cases creates warnings that are misleading or unhelpful.

There was some interest in a "conditional-must-use" based on type parameters, e.g. `Box<T>`/`Option<T>` could be `#[must_use]` iff `T` is `#[must_use]`, but extending this behavior to user-defined types would require an RFC. A `#[rustc_conditional_must_use]` or similar could be appropriate if we want to limit this to std types.

**Comment by eddyb** [comment]
@cramertj "conditional `#[must_use]`" sounds like `#[must_use]` on data types should've been a trait with an associated `const` (for the reason), since then you can just rely on bounds.

I think we can still do that with an unstable trait, and hook it up either through the trait system, or macro expansion, so both `#[must_use]` and the trait work with the lint.
